brokers/artifacts/extensions.go

```diff
@@ -0,0 +1,148 @@
+// installed, returning a list of InstalledPlugins that need to be downloaded. Note that
+// returned list may be partially filled, if e.g. only one of two extensions in a plugin
+// need to be updated.
+// Plugins that are out of date will be deleted from the local filesystem, and any extensions
+// that are not included in the current requested plugins will be removed.
+func (b *Broker) SyncWithPluginsDir(requested []model.CachedPlugin) (toInstall []model.CachedPlugin) {
+	installed, err := b.readInstalledPlugins()
+	if err != nil {
+		// Unable to read -- default to wiping everything and installing all plugins
+		b.PrintDebug("Error encountered during reading installed plugins: %s", err)
+		b.PrintInfo("Failed to get installed plugins")
+		b.resetPluginsDirectory()
+		return requested
+	}
+	toInstall = b.preparePluginsToInstall(requested, installed)
+	return toInstall
+}
+
+// preparePluginsToInstall will prepare the /plugins directory for installation and return a list of (partially-filled)
+// plugins to be installed. Any plugins that are currently installed but not requested in the workspace will
+// be removed from the filesystem.
+func (b *Broker) preparePluginsToInstall(requested, installed []model.CachedPlugin) (toInstall []model.CachedPlugin) {
+	toInstall = requested
+	for _, plugin := range installed {
+		match := findPlugin(plugin, requested)
+		if match == nil {
+			// Plugin has been uninstalled since last start
+			b.PrintInfo("Uninstalling plugin: %s", plugin.ID)
+			if err := b.removePlugin(plugin); err != nil {
+				b.PrintInfo("WARN: failed to remove plugin artifacts: %s", err)
+			}
+			continue
+		}
+		for ext, path := range plugin.CachedExtensions {
+			if _, ok := match.CachedExtensions[ext]; ok {
+				// Extension is already downloaded, fill path in struct to avoid downloading later.
+				match.CachedExtensions[ext] = path
+			} else {
+				// Downloaded plugin is not used in current workspace and must be removed.
+				err := b.ioUtils.RemoveFile(path)
+				if err != nil {
+					b.PrintInfo("WARN: Failed to clean up plugin at %s: %s", path, err)
+				}
+			}
+		}
+	}
+	return toInstall
+}
+
+func (b *Broker) readInstalledPlugins() ([]model.CachedPlugin, error) {
+	bytes, err := b.ioUtils.ReadFile(installedPluginsJSONFile)
+	if err != nil {
+		return nil, err
+	}
+	err = b.ioUtils.RemoveFile(installedPluginsJSONFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var pluginsJSON model.InstalledPluginJSON
+	err = json.Unmarshal(bytes, &pluginsJSON)
+	if err != nil {
+		return nil, err
+	}
+	if pluginsJSON.Version != installedPluginsJSONVersion {
+		b.PrintInfo("Installed plugins cache is incompatible with current version of plugin broker. All plugins will be redownloaded.")
+		return nil, fmt.Errorf("Installed plugins list is from previous version of broker")
+	}
+	return pluginsJSON.Plugins, nil
+}
+
+func (b *Broker) writeInstalledPlugins(cached []model.CachedPlugin) error {
```

Comment:
It's used outside of this file, maybe then it should be public as `SyncWithPluginsDir` is?